### PR TITLE
Add socket support phpfpm

### PR DESCRIPTION
**Pull request by glesys-andreas**
Hi, 

We would like to add option to talk directly to php fpm socket and bypass Apache or Nginx

The module check if the job contains a valid socket och in that case initiates a socket connection instead of a http client.

I appreciate any feedback as it is my first contribution and are a not sure how you want the code formatted and error handling and so on :)

### CODE REVIEW

**Comment by ilyam8** [comment]
@glesys-andreas i see it is still `wip`, i converted the PR to draft. Mark it ready for review when it is ready 😄

**Comment by ilyam8** [comment]
@glesys-andreas 

i think we need to do something like `provider` or `client` interface that returns `status`.

<details>
<summary>status</summary>

```go
type (
	status struct {
		Active    int64  `json:"active processes" stm:"active"`
		MaxActive int64  `json:"max active processes" stm:"maxActive"`
		Idle      int64  `json:"idle processes" stm:"idle"`
		Requests  int64  `json:"accepted conn" stm:"requests"`
		Reached   int64  `json:"max children reached" stm:"reached"`
		Slow      int64  `json:"slow requests" stm:"slow"`
		Processes []proc `json:"processes"`
	}
	requestDuration int64
	proc            struct {
		PID      int64           `json:"pid"`
		State    string          `json:"state"`
		Duration requestDuration `json:"request duration"`
		CPU      float64         `json:"last request cpu"`
		Memory   int64           `json:"last request memory"`
	}
)
```

</details>

```go
type client interface {
  getStatus() (*status, error) 
}

type httpClient struct {
 ...
}

func (c *httpClient) getStatus() (*status, error) {
 ...
}

type socketClient struct {
 ...
}

func (c *socketClient) getStatus() (*status, error) {
 ...
}
```

and init `httpClient` or `socketClient` based on configuration (in `Init()`)

```go
func newClient(conf Config) (client, error) {
 ...
}
```

in that case we dont need to change `Collect()` or `collect()`

**Comment by glesys-andreas** [comment]
@ilyam8 
Sound promising, i like that idea. Can you please elaborate on that, we are not sure if you want a larger rewrite of the module or if this can be inserted in the original code ?

**Comment by ilyam8** [comment]
@glesys-andreas i think i refactor phpfpm collector and then you add unix support. Deal? 🤝

**Comment by glesys-andreas** [comment]
@ilyam8 Sure! Sounds like good deal! :)

**Comment by kiranos** [comment]
@ilyam8 when do you think you will have time to review this?

**Comment by ilyam8** [comment]
@glesys-andreas hey

as soon as i finish https://github.com/netdata/go.d.plugin/pull/421

i hope to finish it today

File: modules/phpfpm/client.go [hunk 1]
```diff
@@ -40,34 +43,69 @@ func (rd *requestDuration) UnmarshalJSON(b []byte) error {
...
 }
 
-type client struct {
+type client interface {
+	getStatus() (*status, error)
+
+}
+
+type httpClient struct {
+	module.Base
```
Comment (ilyam8) [inline modules/phpfpm/client.go:52]: both `httpClient` and `socketClient` should implement `client` only, then dont need to have `module.Base`

`Phpfpm` should have `client` interface

```go
	Phpfpm struct {
		module.Base
		Config `yaml:",inline"`

        client client
	}
```

and we need to initialize appropriate client base on provided configuration in `Init()`

File: modules/phpfpm/client.go [hunk 2]
```diff
@@ -88,3 +126,34 @@ func (c client) fetchStatus(req *http.Request) (*status, error) {
 
 	return s, nil
 }
+
+func (c socketClient) Status() (*status, error) {
+	socket, err := fcgiclient.Dial("unix", c.socket)
```
Comment (ilyam8) [inline modules/phpfpm/client.go:131]: does this library support timeout?

**Review by ilyam8 — CHANGES_REQUESTED** [review]

**Comment by ilyam8** [comment]
Btw i see a lot of ci checks are failed.

File: modules/phpfpm/client.go (hunk 2 above)
Comment (glesys-andreas) [reply modules/phpfpm/client.go:131]: Yes, by Using the function DialTimeout acts like Dial but takes timeout as a third argument.

**Comment by glesys-andreas** [comment]
Addressing the ci checks now

File: modules/phpfpm/client.go (hunk 2 above)
Comment (ilyam8) [reply modules/phpfpm/client.go]: lets use it then

File: modules/phpfpm/client.go [hunk 3]
```diff
@@ -128,7 +121,8 @@ func (c httpClient) fetchStatus(req *http.Request) (*status, error) {
 }
 
 func (c socketClient) Status() (*status, error) {
-	socket, err := fcgiclient.Dial("unix", c.socket)
+	var timeout = time.Duration(1) * 100000000
```
Comment (ilyam8) [inline modules/phpfpm/client.go:124]: there is user configurable timeout

https://github.com/netdata/go.d.plugin/blob/5982f13329a41a0f1235afa9189436e196ae2c9c/modules/phpfpm/phpfpm.go#L24

Comment (glesys-andreas) [reply modules/phpfpm/client.go:124]: Sorry, totaly missed that. i'll add that in the next commit! Thanks!

Comment (ilyam8) [reply modules/phpfpm/client.go:124]: it is `timeout` option in the config

https://github.com/netdata/go.d.plugin/blob/5982f13329a41a0f1235afa9189436e196ae2c9c/config/go.d/phpfpm.conf#L89-L92

File: modules/phpfpm/client.go [hunk 4]
```diff
@@ -88,3 +109,34 @@ func (c client) fetchStatus(req *http.Request) (*status, error) {
 
 	return s, nil
 }
+
+func (c socketClient) Status() (*status, error) {
+	socket, err := fcgiclient.DialTimeout("unix", c.socket, c.timeout)
+	if err != nil {
+		return nil, fmt.Errorf("error on connecting to socket: %v", err)
+	}
+
```
Comment (ilyam8) [inline modules/phpfpm/client.go:118]: i think we need to `defer socket.Close()` right after `fcgiclient.DialTimeout` if there is no error

File: modules/phpfpm/client.go [hunk 5]
```diff
@@ -88,3 +109,34 @@ func (c client) fetchStatus(req *http.Request) (*status, error) {
...
+	return c.fetchStatus(socket)
+}
+
+func (c socketClient) fetchStatus(socket *fcgiclient.FCGIClient) (*status, error) {
+
+	resp, err := socket.Get(c.env)
```
Comment (ilyam8) [inline modules/phpfpm/client.go:124]: empty line `123`

File: modules/phpfpm/client.go [hunk 6]
```diff
@@ -88,3 +109,34 @@ func (c client) fetchStatus(req *http.Request) (*status, error) {
...
+
+	content, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error on reading socket: %v", err)
+	}
+
+	socket.Close()
+	st := &status{}
+
+	err2 := json.Unmarshal(content, st)
```
Comment (ilyam8) [inline modules/phpfpm/client.go:137]: `err = json.Unmarshal(content, st)`, no need for `err2`

**Comment by ilyam8** [comment]
@glesys-andreas i think this is almost done, there are some changes i would like but they dont affect functionality so i can do them later myself

**Comment by ilyam8** [comment]
ok, i am going to release new version today, so am merging this PR and doing some minor changes in a follow up PR

**Review by ilyam8 — APPROVED** [review]

**Pull request merged**